File: app/src/main/java/com/mlsdev/mlsdevstore/data/local/database/dao/CategoriesDao.java
package com.mlsdev.mlsdevstore.data.local.database.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import com.mlsdev.mlsdevstore.data.local.database.Table;
import com.mlsdev.mlsdevstore.data.model.category.Category;
import com.mlsdev.mlsdevstore.data.model.category.CategoryTree;
import com.mlsdev.mlsdevstore.data.model.category.CategoryTreeNode;

import java.util.List;

import io.reactivex.Single;

@Dao
public interface CategoriesDao {

    @Insert
    void insertCategories(Category... category);

    @Update
    void updateCategories(Category... categories);

    @Delete
    void deleteCategories(Category... categories);

    @Query("delete from " + Table.CATEGORY_TREE_NODES)
    int deleteAllCategoryTreeNodel();

    @Query("select * from " + Table.CATEGORIES)
    Single<List<Category>> queryAllCategories();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertCategoryTreeNode(CategoryTreeNode... categoryTreeNodes);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertCategoryTreeNode(List<CategoryTreeNode> categoryTreeNodes);

    @Query("select * from " + Table.CATEGORY_TREE_NODES)
    Single<List<CategoryTreeNode>> queryCategoryTreeNode();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertCategoryTree(CategoryTree categoryTree);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertCategoryTree(CategoryTree... categoryTree);

    @Query("select * from " + Table.CATEGORY_TREES + " limit 1")
    Single<List<CategoryTree>> queryDefaultCategoryTree();

    @Query("delete from " + Table.CATEGORY_TREES)
    int deleteAllCategoryTrees();

}

File: app/src/main/java/com/mlsdev/mlsdevstore/data/local/database/dao/CategoriesDao.kt
package com.mlsdev.mlsdevstore.data.local.database.dao

import androidx.room.*
import com.mlsdev.mlsdevstore.data.local.database.Table
import com.mlsdev.mlsdevstore.data.model.category.Category
import com.mlsdev.mlsdevstore.data.model.category.CategoryTree
import com.mlsdev.mlsdevstore.data.model.category.CategoryTreeNode
import io.reactivex.Single

@Dao
interface CategoriesDao {

    @Insert
    fun insertCategories(vararg category: Category)

    @Update
    fun updateCategories(vararg categories: Category)

    @Delete
    fun deleteCategories(vararg categories: Category)

    @Query("delete from ${Table.CATEGORY_TREE_NODES}")
    fun deleteAllCategoryTreeNodes(): Int

    @Query("select * from ${Table.CATEGORIES}")
    fun queryAllCategories(): Single<List<Category>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertCategoryTreeNode(vararg categoryTreeNodes: CategoryTreeNode)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertCategoryTreeNode(categoryTreeNodes: List<CategoryTreeNode>)

    @Query("select * from ${Table.CATEGORY_TREE_NODES}")
    fun queryCategoryTreeNode(): Single<List<CategoryTreeNode>>

    @Query("select * from ${Table.CATEGORY_TREE_NODES}")
    fun queryCategoryTreeNodeSync(): List<CategoryTreeNode>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertCategoryTree(categoryTree: CategoryTree)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertCategoryTree(vararg categoryTree: CategoryTree)

    @Query("select * from ${Table.CATEGORY_TREES} limit 1")
    fun queryDefaultCategoryTree(): Single<List<CategoryTree>>

    @Query("delete from ${Table.CATEGORY_TREES}")
    fun deleteAllCategoryTrees(): Int

}
